go/vt/tabletmanager/rpc_replication.go

```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
 	return false
 }
 
-func (agent *ActionAgent) enableSemiSync(master bool) error {
+func (agent *ActionAgent) setSemiSync(tabletType topodatapb.TabletType) error {
+	if !*enableSemiSync {
+		// Semi-sync handling is not enabled.
+		return nil
+	}
+
 	// Only enable if we're eligible for becoming master (REPLICA type).
 	// Ineligible slaves (RDONLY) shouldn't ACK because we'll never promote them.
-	if !isMasterEligible(agent.Tablet().Type) {
-		return nil
+	if !isMasterEligible(tabletType) {
+		return agent.MysqlDaemon.SetSemiSyncEnabled(false, false)
 	}
 
 	// Always enable slave-side since it doesn't hurt to keep it on for a master.
 	// The master-side needs to be off for a slave, or else it will get stuck.
-	return agent.MysqlDaemon.SetSemiSyncEnabled(master, true)
+	return agent.MysqlDaemon.SetSemiSyncEnabled(tabletType == topodatapb.TabletType_MASTER, true)
+}
+
+func (agent *ActionAgent) setSemiSyncAndFixReplication(tabletType topodatapb.TabletType) error {
+	if !*enableSemiSync {
+		// Semi-sync handling is not enabled.
+		return nil
+	}
+
+	if tabletType == topodatapb.TabletType_MASTER {
+		// Master is special. It is always handled at the
+		// right time by the reparent operations, it doesn't
+		// need to be fixed.
+		return nil
+	}
+
+	if err := agent.setSemiSync(tabletType); err != nil {
+		return fmt.Errorf("failed to setSemiSync(%v): %v", tabletType, err)
+	}
+
+	// If replication is running, but the status is wrong,
+	// restart replication.
+	if _, err := agent.MysqlDaemon.SlaveStatus(); err != nil {
+		// Replication is not running, nothing to do.
+		return nil
+	}
+
+	shouldAck := isMasterEligible(tabletType)
+	acking, err := agent.MysqlDaemon.SemiSyncSlaveStatus()
+	if err != nil {
+		return fmt.Errorf("failed to get SemiSyncSlaveStatus: %v", err)
```

Comment:
Next I also want to call this function in more places, like in RefreshState. So with this logic, it's better, we don't stop/start replication when we don't need to.

With the error case, I'd rather stop asap. If we can't get the flag, it means we can't reliably connect to mysql, I'd rather not touch it more.